src/librustc_middle/ty/instance.rs

```diff
@@ -119,12 +120,6 @@ impl<'tcx> Instance<'tcx> {
     /// the `ParamEnv` for our generic context.
     pub fn monomorphic_ty(&self, tcx: TyCtxt<'tcx>) -> Ty<'tcx> {
         let ty = tcx.type_of(self.def.def_id());
-        // There shouldn't be any params - if there are, then
-        // Instance.ty_env should have been used to provide the proper
-        // ParamEnv
-        if self.substs.has_param_types_or_consts() {
-            bug!("Instance.ty called for type {:?} with params in substs: {:?}", ty, self.substs);
-        }
```

Comment:
The comment on this method and the name of the method are still not great.

IMO you should rename  `ty_env` to `ty` and just pass `ty::ParamEnv::reveal_all()` from `monomorphic_ty()` call sites.
You'll also need to merge their doc comments.

(eventually codegen will need to keep track of a `ParamEnv` instead of assuming `ty::ParamEnv::reveal_all()` everywhere, in order to make even basic layout-sensitive polymorphization work, due to `T: Sized` bounds being needed)

Reply:
Fixed!